sycl/source/detail/plugin.hpp

```diff
@@ -184,11 +184,67 @@ class plugin {
   void *getLibraryHandle() { return MLibraryHandle; }
   int unload() { return RT::unloadPlugin(MLibraryHandle); }
 
+  // return the index of PiPlatforms.
+  // If not found, add it and return its index.
+  int getPlatformId(RT::PiPlatform Platform) {
+    if (PiPlatforms) {
+      auto It = std::find(PiPlatforms->begin(), PiPlatforms->end(), Platform);
+      if (It != PiPlatforms->end())
+        return It - PiPlatforms->begin();
+
+      PiPlatforms->push_back(Platform);
+      LastDeviceIds.push_back(0);
+      return PiPlatforms->size() - 1;
+    }
+    return -1;
+  }
+  // Device ids are consecutive across platforms within a plugin.
+  // We need to return the same starting index for the given platform.
+  // So, instead of returing the last device id of the given platform,
+  // return the last device id of the predecessor platform.
+  int getStartingDeviceId(RT::PiPlatform Platform) {
+    if (!PiPlatforms)
+      PiPlatforms = new std::vector<RT::PiPlatform>;
+    int PlatformId = getPlatformId(Platform);
+    if (PlatformId <= 0)
+      return 0;
+
+    return LastDeviceIds[PlatformId - 1];
+  }
+  // set the id of the last device for the given platform
+  void setLastDeviceId(RT::PiPlatform Platform, int Id) {
+    int PlatformId = getPlatformId(Platform);
+    LastDeviceIds[PlatformId] = Id;
+  }
+  // reset all last device ids to zeros
+  void resetLastDeviceIds() {
+    std::fill(LastDeviceIds.begin(), LastDeviceIds.end(), 0);
+  }
+
+  void resetPiPlatforms() {
+    delete PiPlatforms;
+    PiPlatforms = nullptr;
+  }
+
+  bool containsPiPlatform(RT::PiPlatform Platform) {
+    if (PiPlatforms) {
+      auto It = std::find(PiPlatforms->begin(), PiPlatforms->end(), Platform);
+      if (It != PiPlatforms->end())
+        return true;
+    }
+    return false;
+  }
+
 private:
   RT::PiPlugin MPlugin;
   backend MBackend;
   void *MLibraryHandle; // the handle returned from dlopen
   std::shared_ptr<std::mutex> TracingMutex;
+  // vector of PiPlatforms that belong to this plugin
+  std::vector<RT::PiPlatform> *PiPlatforms = nullptr;
```

Comment:
Please, replace with just `std::vector`(not a pointer to `std::vector`) or clarify why it cannot be done.

Reply:
When we do shutdown(), we destroy all global plugins.
Each plugin contains a vector of PiPlatforms.
PiPlatforms are pointers to pi_platform objects.
So, the destructor of the PiPlatforms vector will call the destructor for each individual element.
This causes pi_platform objects to be destroyed twice, once from RT::shutdown() and the other from plugin::shutdown().

Reply:
>PiPlatforms are pointers to pi_platform objects.
> So, the destructor of the PiPlatforms vector will call the destructor for each individual element.

In C++ a destructor for an object, which a pointer points to, is not called automatically. So,
```
{
  int *Array = new int[1]; // memory allocated
  std::vector<int *> VecOfInts{1};
  VecOfInts[0] = Array;
} // VecOfInts is destroyed
```
the dynamic memory which `Array` points to is not deallocated when `VecOfInts` is destroyed.

Sorry, if I'm missing something obvious here, but it's still not clear to me.

Reply:
Actually, my case is different.
The `plugin` class contains a vector of pointers to another object.
<pre>
class plugin {
   std::vector<pi_platform*> PiPlatforms;
}
</pre>
When a plugin object is destroyed, the data member `PiPlatforms` is also destroyed automatically.
Since this data member is a vector, C++ realizes that the element is a pointer to another class.
So, it invokes destructor for each element in the `PiPlatforms` vector.
Note that plugins also maintain the list of pi_platform pointers to deallocate memory that the plugin allocated.
But, SYCL RT already destroyed memory for the pi_platform object, and it is going to crash because the plugin does not know whether RT already destroyed it as a part of the shutdown process.

Reply:
>> Since this data member is a vector, C++ realizes that the element is a pointer to another class.
>> So, it invokes destructor for each element in the PiPlatforms vector.

Right, it calls desctructor for each element, but the element type here is a pointer with is "builtin" type(not a class) and has no  C++ destructor. So, during PiPlatforms destruction the only thing which will happen is deallocation of the memory the vector manages.

Furthemore there is no way to call `delete` on a variable of `RT::PiPlatform*` type since the PiPlatform is an incomplete object.

Reply:
I think you are right.
But, somehow PiPlatform* were destroyed already when I did not reset them before RT performs shutdown().
Let me try to skip calling resetPiPlatforms() in RT shutdown.

Reply:
I agree.
I think I need a separate lock for DeviceId and PlatformId within a plugin class.

Reply:
Done. I changed to a vector data member instead of a pointer to the vector.